src/EditorFeatures/CSharpTest/ForeachToFor/ForeachToForTests.cs

```diff
@@ -0,0 +1,1248 @@
+    public int Count => throw new NotImplementedException();
+    public IEnumerator<int> GetEnumerator() => throw new NotImplementedException();
+    IEnumerator IEnumerable.GetEnumerator() => throw new NotImplementedException();
+
+    string IReadOnlyList<string>.this[int index] => throw new NotImplementedException();
+    int IReadOnlyCollection<string>.Count => throw new NotImplementedException();
+    IEnumerator<string> IEnumerable<string>.GetEnumerator() => throw new NotImplementedException();
+}
+";
+            await TestInRegularAndScriptAsync(text, expected);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.ForeachToFor)]
+        public async Task PreserveUserExpression()
+        {
+            var text = @"
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace NS
+{
+    class Test
+    {
+        void Method()
+        {
+            foreach [||] (string a in new NS.Mixed())
+            {
+                Console.WriteLine(a);
+            }
+        }
+    }
+
+    class Mixed : IReadOnlyList<int>, IReadOnlyList<string>
+    {
+        public int this[int index] => throw new NotImplementedException();
+        public int Count => throw new NotImplementedException();
+        public IEnumerator<int> GetEnumerator() => throw new NotImplementedException();
+        IEnumerator IEnumerable.GetEnumerator() => throw new NotImplementedException();
+
+        string IReadOnlyList<string>.this[int index] => throw new NotImplementedException();
+        int IReadOnlyCollection<string>.Count => throw new NotImplementedException();
+        IEnumerator<string> IEnumerable<string>.GetEnumerator() => throw new NotImplementedException();
+    }
+}
+";
+
+            var expected = @"
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace NS
+{
+    class Test
+    {
+        void Method()
+        {
+            var {|Rename:list|} = (IReadOnlyList<string>)new NS.Mixed();
+            for (var {|Rename:i|} = 0; i < list.Count; i++)
+            {
+                var a = list[i];
+                Console.WriteLine(a);
+            }
+        }
+    }
+
+    class Mixed : IReadOnlyList<int>, IReadOnlyList<string>
+    {
+        public int this[int index] => throw new NotImplementedException();
+        public int Count => throw new NotImplementedException();
+        public IEnumerator<int> GetEnumerator() => throw new NotImplementedException();
+        IEnumerator IEnumerable.GetEnumerator() => throw new NotImplementedException();
+
+        string IReadOnlyList<string>.this[int index] => throw new NotImplementedException();
+        int IReadOnlyCollection<string>.Count => throw new NotImplementedException();
+        IEnumerator<string> IEnumerable<string>.GetEnumerator() => throw new NotImplementedException();
+    }
+}
+";
+            await TestInRegularAndScriptAsync(text, expected);
+        }
+    }
+}
```

Comment:
Should also test with deconstruction-foreach: `foreach (var (x, y) in collection) ...`
It's possible it's a different syntax node, in which case there is no interaction with refactoring.